core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala

```diff
@@ -617,26 +615,88 @@ class SpendReportingService(
           val workspaceNamesByProjectId: Map[GoogleProjectId, WorkspaceName] = billingMapForQuery.values.flatten
             .map(ws => ws.googleProjectId -> ws.toWorkspaceName)
             .toMap
-
-          val query = getAllUserWorkspaceQuery(tableNameForQuery, billingProjectsByAccount, pageSize, offset)
-          val queryJob = setUpAllUserWorkspaceQuery(query, start, end)
-          runBigQueryJob(queryJob, childContext)
-            .map { result =>
-              result.getValues.asScala.toList match {
-                case Nil => None
-                case rows =>
-                  Some(extractCrossBillingProjectSpendReportingResults(rows, start, end, workspaceNamesByProjectId))
+          val projectIds = workspaceNamesByProjectId.keySet.map(_.value)
+          val cachedResults =
+            workspaceSpendReportRepository.getWorkspaceSpendReports(projectIds, startLocalDateTime, endLocalDateTime)
+          val spendResults = cachedResults.flatMap { cachedResult =>
+            if (cachedResult.size == projectIds.size) {
+              val hasDataAvailable = cachedResult.filter(cached => cached.isDataAvailable)
+              if (hasDataAvailable.nonEmpty) {
+                val spendReportingResults =
+                  WorkspaceSpendReportRecord.toSpendReportingResults(hasDataAvailable, workspaceNamesByProjectId)
+                Future.successful(
+                  Some(spendReportingResults)
+                )
+              } else {
+                Future.successful(None)
               }
+            } else {
+              val query = getAllUserWorkspaceQuery(tableNameForQuery, billingProjectsByAccount, pageSize, offset)
+              val queryJob = setUpAllUserWorkspaceQuery(query, start, end)
+              val queryResults = runBigQueryJob(queryJob, childContext)
+                .map { result =>
+                  result.getValues.asScala.toList match {
+                    case Nil => None
+                    case rows =>
+                      val crossBillingResults =
+                        extractCrossBillingProjectSpendReportingResults(rows, start, end, workspaceNamesByProjectId)
+                      workspaceSpendReportRepository.insertSpendReportResults(crossBillingResults)
+                      Some(crossBillingResults)
+                  }
+                }
+                .recoverWith { case ex: Throwable =>
+                  logger.warn(s"Error fetching results from BigQuery: ${ex.getMessage}")
+                  Future.successful(None)
+                }
+              queryResults.flatMap { res =>
+                Future.successful(insertRecordsWithMissingSpendData(res, workspaceNamesByProjectId, start, end))
+              }
```

Comment:
Two suggestions here:

1. you can simplify this with `.map` instead of `.flatMap`:
```suggestion
              queryResults.map { res =>
                insertRecordsWithMissingSpendData(res, workspaceNamesByProjectId, start, end)
              }

```

2. I think this should catch and ignore any errors that happen when saving records to the cache. If we hit an error here, we can still deliver results to the end user; we'll just fail to cache those results. There is a race condition which makes errors here likely: since the calls to `workspaceSpendReportRepository.getWorkspaceSpendReports` and `workspaceSpendReportRepository.insertWorkspaceSpendReport` are each in their own separate transaction, if multiple users hit an uncached consolidated spend report at the same time, both could see no records in cache, and then both would try to write those records to cache. One of those writes will fail due to the unique constraints on the cache table.